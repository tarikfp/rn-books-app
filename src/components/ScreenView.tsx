import omit from 'lodash/omit';
import * as React from 'react';
import {
  Animated as RNAnimated,
  Keyboard,
  ScrollView,
  ScrollViewProps,
  StatusBar,
  StatusBarProps,
  TouchableWithoutFeedback,
  TouchableWithoutFeedbackProps,
} from 'react-native';
import Animated from 'react-native-reanimated';
import {
  SafeAreaView,
  SafeAreaViewProps,
} from 'react-native-safe-area-context';
import { s } from '~theme/common-styles';
import { AppTheme, useAppTheme, useThemePreference } from '~theme/core';
import { Box } from './Box';

type Props = {
  withSafeArea?: boolean;
  withStatusBar?: boolean;
  withTouchableWithoutFeedback?: boolean;
  withScrollView?: boolean;
  shouldDismissKeyboardOnPressOutside?: boolean;
  statusBarProps?: StatusBarProps;
  backgroundColor?: keyof AppTheme['colors'];
  rootViewProps?: React.ComponentProps<typeof Box>;
  touchableWithoutFeedbackProps?: TouchableWithoutFeedbackProps;
  safeAreaProps?: SafeAreaViewProps;
  scrollViewProps?: ScrollViewProps & {
    animatedScrollViewProps?: Animated.AnimateProps<ScrollViewProps>;
    rnAnimatedProps?: React.ComponentProps<typeof Animated.ScrollView>;
  };
  disableHorizontalPadding?: boolean;
  disableVerticalPadding?: boolean;
  padding?: keyof AppTheme['spacing'];
};

interface WithComponentPropsBase<P, T = undefined> {
  shouldRender: boolean;
  extraProps?: P;
  ref?: React.ForwardedRef<T>;
}

type WithScrollViewProps = WithComponentPropsBase<
  ScrollViewProps & {
    animatedScrollViewProps?: Animated.AnimateProps<ScrollViewProps>;
    rnAnimatedProps?: React.ComponentProps<typeof Animated.ScrollView>;
  },
  ScrollView
>;
type WithTouchableWithoutFeedbackProps = WithComponentPropsBase<
  React.ComponentProps<typeof TouchableWithoutFeedback>
>;
type WithSafeAreaProps = WithComponentPropsBase<SafeAreaViewProps>;
type WithStatusBarProps = WithComponentPropsBase<StatusBarProps>;

const WithTouchableWithoutFeedback: React.FC<
  React.PropsWithChildren<WithTouchableWithoutFeedbackProps>
> = ({ children, extraProps, shouldRender }) => {
  if (!shouldRender) {
    return children as React.ReactElement;
  }

  return (
    <TouchableWithoutFeedback
      style={[s.flex(), extraProps?.style]}
      {...omit(extraProps, 'style')}
    >
      {children}
    </TouchableWithoutFeedback>
  );
};

const WithScrollView: React.FC<
  React.PropsWithChildren<WithScrollViewProps>
> = ({ extraProps, shouldRender, children, ref }) => {
  if (!shouldRender) {
    return children as React.ReactElement;
  }

  /** TODO: improve more on converting a component to a animated component, make it more generic for other components */

  if (typeof extraProps?.animatedScrollViewProps !== 'undefined') {
    return (
      <Animated.ScrollView
        ref={ref as React.ForwardedRef<Animated.ScrollView>}
        {...extraProps.animatedScrollViewProps}
      >
        {children}
      </Animated.ScrollView>
    );
  }

  if (typeof extraProps?.rnAnimatedProps !== 'undefined') {
    return (
      <RNAnimated.ScrollView
        ref={ref as React.ForwardedRef<Animated.ScrollView>}
        {...extraProps.rnAnimatedProps}
      >
        {children}
      </RNAnimated.ScrollView>
    );
  }

  return (
    <ScrollView ref={ref} {...omit(extraProps, 'animatedScrollViewProps')}>
      {children}
    </ScrollView>
  );
};

const WithSafeArea: React.FC<React.PropsWithChildren<WithSafeAreaProps>> = ({
  children,
  extraProps,
  shouldRender,
}) => {
  const { colors } = useAppTheme();
  if (!shouldRender) {
    return children as React.ReactElement;
  }

  return (
    <SafeAreaView
      style={[
        s.flex(),
        { backgroundColor: colors.background },
        extraProps?.style,
      ]}
      {...omit(extraProps, 'style')}
    >
      {children}
    </SafeAreaView>
  );
};

const WithAppStatusBar: React.FC<WithStatusBarProps> = ({
  shouldRender,
  extraProps,
}) => {
  const { isDarkTheme } = useThemePreference();

  if (!shouldRender) {
    return null;
  }

  return (
    <StatusBar
      barStyle={isDarkTheme ? 'light-content' : 'dark-content'}
      {...extraProps}
    />
  );
};

const WithRootView: React.FC<
  React.PropsWithChildren<React.ComponentProps<typeof Box>>
> = ({ children, ...props }) => {
  return <Box {...props}>{children}</Box>;
};

export const ScreenView: React.FC<React.PropsWithChildren<Props>> = ({
  backgroundColor,
  children,
  scrollViewProps,
  rootViewProps,
  statusBarProps,
  safeAreaProps,
  touchableWithoutFeedbackProps,
  shouldDismissKeyboardOnPressOutside = false,
  disableHorizontalPadding = false,
  disableVerticalPadding = false,
  withSafeArea = false,
  withTouchableWithoutFeedback = false,
  withStatusBar = false,
  withScrollView = false,
  padding,
}) => {
  const computedBgColor = backgroundColor ?? 'background';

  const computedPadding: keyof AppTheme['spacing'] = padding ?? 'm';

  return (
    <WithSafeArea shouldRender={withSafeArea} extraProps={safeAreaProps}>
      <WithAppStatusBar shouldRender={withStatusBar} {...statusBarProps} />
      <WithScrollView
        shouldRender={withScrollView}
        extraProps={scrollViewProps}
      >
        <WithRootView
          flex={1}
          backgroundColor={computedBgColor}
          paddingVertical={disableVerticalPadding ? undefined : computedPadding}
          paddingHorizontal={
            disableHorizontalPadding ? undefined : computedPadding
          }
          {...rootViewProps}
        >
          <WithTouchableWithoutFeedback
            extraProps={{
              onPress: shouldDismissKeyboardOnPressOutside
                ? Keyboard.dismiss
                : undefined,
              ...touchableWithoutFeedbackProps,
            }}
            shouldRender={
              withTouchableWithoutFeedback ||
              shouldDismissKeyboardOnPressOutside
            }
          >
            {children}
          </WithTouchableWithoutFeedback>
        </WithRootView>
      </WithScrollView>
    </WithSafeArea>
  );
};
