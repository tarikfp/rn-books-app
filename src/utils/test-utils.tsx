import { ThemeProvider } from '@shopify/restyle';
import { render, RenderOptions } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { theme } from '~theme/core';
import { ThemePreferenceContext } from '~theme/useThemePreference';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemePreferenceContext.Provider
      value={{
        isDarkTheme: false,
        toggleTheme: () => {},
      }}
    >
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </ThemePreferenceContext.Provider>
  );
};

const withAllTheProviders = (ui: React.ReactNode) => (
  <AllTheProviders children={ui} />
);

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderer = (ui: React.ReactElement) =>
  renderer.create(withAllTheProviders(ui));

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { customRender as render };
// override renderer method
export { customRenderer as createRenderer };
