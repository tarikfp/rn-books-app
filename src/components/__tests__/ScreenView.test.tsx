import React from 'react';
import { Text } from 'react-native';
import { render } from '~utils/test-utils';
import { ScreenView } from '../ScreenView';

describe('ScreenView', () => {
  it('renders correctly with default props', () => {
    const { toJSON } = render(
      <ScreenView>
        <Text>Test Content</Text>
      </ScreenView>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with custom background color', () => {
    const { toJSON } = render(
      <ScreenView backgroundColor="primary">
        <Text>Test Content</Text>
      </ScreenView>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with safe area enabled', () => {
    const { toJSON } = render(
      <ScreenView withSafeArea>
        <Text>Test Content</Text>
      </ScreenView>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with status bar enabled', () => {
    const { toJSON } = render(
      <ScreenView withStatusBar>
        <Text>Test Content</Text>
      </ScreenView>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with touchable without feedback enabled', () => {
    const { toJSON } = render(
      <ScreenView withTouchableWithoutFeedback>
        <Text>Test Content</Text>
      </ScreenView>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with scroll view enabled', () => {
    const { toJSON } = render(
      <ScreenView withScrollView>
        <Text>Test Content</Text>
      </ScreenView>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with custom padding', () => {
    const { toJSON } = render(
      <ScreenView padding="l">
        <Text>Test Content</Text>
      </ScreenView>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with all custom props enabled', () => {
    const { toJSON } = render(
      <ScreenView
        backgroundColor="primary"
        withSafeArea
        withStatusBar
        withTouchableWithoutFeedback
        withScrollView
        padding="l"
      >
        <Text>Test Content</Text>
      </ScreenView>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with shouldDismissKeyboardOnPressOutside prop', () => {
    const { toJSON } = render(
      <ScreenView shouldDismissKeyboardOnPressOutside>
        <Text>Test Content</Text>
      </ScreenView>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with disableHorizontalPadding prop', () => {
    const { toJSON } = render(
      <ScreenView disableHorizontalPadding>
        <Text>Test Content</Text>
      </ScreenView>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with disableVerticalPadding prop', () => {
    const { toJSON } = render(
      <ScreenView disableVerticalPadding>
        <Text>Test Content</Text>
      </ScreenView>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with all custom props and edge cases', () => {
    const { toJSON } = render(
      <ScreenView
        backgroundColor="primary"
        withSafeArea
        withStatusBar
        withTouchableWithoutFeedback
        withScrollView
        padding="l"
        shouldDismissKeyboardOnPressOutside
        disableHorizontalPadding
        disableVerticalPadding
      >
        <Text>Test Content</Text>
      </ScreenView>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with all props disabled', () => {
    const { toJSON } = render(
      <ScreenView>
        <Text>Test Content</Text>
      </ScreenView>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
