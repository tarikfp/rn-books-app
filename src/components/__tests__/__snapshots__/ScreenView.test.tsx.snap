// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScreenView renders correctly with all custom props and edge cases 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    [
      {
        "flex": 1,
      },
      {
        "backgroundColor": "#F3F4F6",
      },
      undefined,
    ]
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          [
            {
              "backgroundColor": "#3EBCB9",
              "flex": 1,
            },
          ]
        }
      >
        <Text
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          Test Content
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;

exports[`ScreenView renders correctly with all custom props enabled 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    [
      {
        "flex": 1,
      },
      {
        "backgroundColor": "#F3F4F6",
      },
      undefined,
    ]
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          [
            {
              "backgroundColor": "#3EBCB9",
              "flex": 1,
              "paddingHorizontal": 24,
              "paddingVertical": 24,
            },
          ]
        }
      >
        <Text
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          Test Content
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;

exports[`ScreenView renders correctly with all props disabled 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F3F4F6",
        "flex": 1,
        "paddingHorizontal": 16,
        "paddingVertical": 16,
      },
    ]
  }
>
  <Text>
    Test Content
  </Text>
</View>
`;

exports[`ScreenView renders correctly with custom background color 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#3EBCB9",
        "flex": 1,
        "paddingHorizontal": 16,
        "paddingVertical": 16,
      },
    ]
  }
>
  <Text>
    Test Content
  </Text>
</View>
`;

exports[`ScreenView renders correctly with custom padding 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F3F4F6",
        "flex": 1,
        "paddingHorizontal": 24,
        "paddingVertical": 24,
      },
    ]
  }
>
  <Text>
    Test Content
  </Text>
</View>
`;

exports[`ScreenView renders correctly with default props 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F3F4F6",
        "flex": 1,
        "paddingHorizontal": 16,
        "paddingVertical": 16,
      },
    ]
  }
>
  <Text>
    Test Content
  </Text>
</View>
`;

exports[`ScreenView renders correctly with disableHorizontalPadding prop 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F3F4F6",
        "flex": 1,
        "paddingVertical": 16,
      },
    ]
  }
>
  <Text>
    Test Content
  </Text>
</View>
`;

exports[`ScreenView renders correctly with disableVerticalPadding prop 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F3F4F6",
        "flex": 1,
        "paddingHorizontal": 16,
      },
    ]
  }
>
  <Text>
    Test Content
  </Text>
</View>
`;

exports[`ScreenView renders correctly with safe area enabled 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    [
      {
        "flex": 1,
      },
      {
        "backgroundColor": "#F3F4F6",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#F3F4F6",
          "flex": 1,
          "paddingHorizontal": 16,
          "paddingVertical": 16,
        },
      ]
    }
  >
    <Text>
      Test Content
    </Text>
  </View>
</RNCSafeAreaView>
`;

exports[`ScreenView renders correctly with scroll view enabled 1`] = `
<RCTScrollView>
  <View>
    <View
      style={
        [
          {
            "backgroundColor": "#F3F4F6",
            "flex": 1,
            "paddingHorizontal": 16,
            "paddingVertical": 16,
          },
        ]
      }
    >
      <Text>
        Test Content
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`ScreenView renders correctly with shouldDismissKeyboardOnPressOutside prop 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F3F4F6",
        "flex": 1,
        "paddingHorizontal": 16,
        "paddingVertical": 16,
      },
    ]
  }
>
  <Text
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    Test Content
  </Text>
</View>
`;

exports[`ScreenView renders correctly with status bar enabled 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F3F4F6",
        "flex": 1,
        "paddingHorizontal": 16,
        "paddingVertical": 16,
      },
    ]
  }
>
  <Text>
    Test Content
  </Text>
</View>
`;

exports[`ScreenView renders correctly with touchable without feedback enabled 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F3F4F6",
        "flex": 1,
        "paddingHorizontal": 16,
        "paddingVertical": 16,
      },
    ]
  }
>
  <Text
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    Test Content
  </Text>
</View>
`;
